File: python/p003.py
# Compute the max prime factor of 600,851,475,143
# It is easily verified that 71 is the
# smallest prime divisor, and thus we only need to check
# to sqrt of the number (it is not prime itself)
from math import sqrt
from utils import is_prime, primes


def compute():
    TARGET = 600_851_475_143
    PRIMES = primes(int(sqrt(TARGET)) + 1)
    max_div = 0
    for p in PRIMES:
        if TARGET % p == 0:
            max_div = p
        while TARGET % p == 0:
            TARGET //= p
    if TARGET > max_div:
        return TARGET
    else:
        return max_div


if __name__ == '__main__':
    print(compute())

File: java/p003.java
import java.lang.Math;

public class p003 {

    public static void main(String[] args) {
        System.out.println(p003.run());
    }

    public static long run() {
        long maxDiv = 0;
        long x = 600851475143L;
        for (long i = 2; i < Math.floor(Math.sqrt(x)); i++) {
            if (x % i == 0 && isPrime(i)) {
                maxDiv = i;
            }
        }
        return maxDiv;
    }

    public static boolean isPrime(long n) {
        if (n < 2) {
            return false;
        }
        else if (n == 2 || n == 3) {
            return true;
        }
        else if (n % 2 == 0) {
            return false;
        }
        else {
            for (int i = 3; i < Math.ceil(Math.sqrt(n)) + 1; i++) {
                if (n % i == 0) {
                    return false;
                }
            }
            return true;
        }
    }
}